{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c93f3735-8241-42d2-a970-8736df333f95",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "523e497f-61e3-4de0-ba52-79a0b45d9c02",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c12f7a45-3194-4f04-84bb-977c90f46472",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3214482</td>\n",
       "      <td>3216968</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>3216022</td>\n",
       "      <td>3216024</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>3216025</td>\n",
       "      <td>3216968</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>3421702</td>\n",
       "      <td>3421901</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3421702</td>\n",
       "      <td>3421901</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0        1           2        3        4  5  6  7  \\\n",
       "0  chr1  unknown        exon  3214482  3216968  .  -  .   \n",
       "1  chr1  unknown  stop_codon  3216022  3216024  .  -  .   \n",
       "2  chr1  unknown         CDS  3216025  3216968  .  -  2   \n",
       "3  chr1  unknown         CDS  3421702  3421901  .  -  1   \n",
       "4  chr1  unknown        exon  3421702  3421901  .  -  .   \n",
       "\n",
       "                                                   8  \n",
       "0  gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  \n",
       "1  gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  \n",
       "2  gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  \n",
       "3  gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  \n",
       "4  gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf = pd.read_csv(\"/Users/lihsinchang/Desktop/genes.gtf\", comment = \"#\", sep = \"\\t\", header=None)\n",
    "gtf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "879a16ae-ff5c-4b7d-94fa-5a0ae2499256",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attritubes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3214482</td>\n",
       "      <td>3216968</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>3216022</td>\n",
       "      <td>3216024</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>3216025</td>\n",
       "      <td>3216968</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>3421702</td>\n",
       "      <td>3421901</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3421702</td>\n",
       "      <td>3421901</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr   source     feature    start      end score strand frame  \\\n",
       "0  chr1  unknown        exon  3214482  3216968     .      -     .   \n",
       "1  chr1  unknown  stop_codon  3216022  3216024     .      -     .   \n",
       "2  chr1  unknown         CDS  3216025  3216968     .      -     2   \n",
       "3  chr1  unknown         CDS  3421702  3421901     .      -     1   \n",
       "4  chr1  unknown        exon  3421702  3421901     .      -     .   \n",
       "\n",
       "                                          attritubes  \n",
       "0  gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  \n",
       "1  gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  \n",
       "2  gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  \n",
       "3  gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  \n",
       "4  gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  "
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#3. Assign column names\n",
    "gtf.columns = [\"chr\", \"source\", \"feature\", \"start\", \"end\", \"score\", \"strand\", \"frame\", \"attritubes\"] \n",
    "gtf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "dcc09577-c345-44a0-a774-e7136b1b2e40",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attritubes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3214482</td>\n",
       "      <td>3216968</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>3216022</td>\n",
       "      <td>3216024</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>3216025</td>\n",
       "      <td>3216968</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>3421702</td>\n",
       "      <td>3421901</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3421702</td>\n",
       "      <td>3421901</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr   source     feature    start      end score strand frame  \\\n",
       "0  chr1  unknown        exon  3214482  3216968     .      -     .   \n",
       "1  chr1  unknown  stop_codon  3216022  3216024     .      -     .   \n",
       "2  chr1  unknown         CDS  3216025  3216968     .      -     2   \n",
       "3  chr1  unknown         CDS  3421702  3421901     .      -     1   \n",
       "4  chr1  unknown        exon  3421702  3421901     .      -     .   \n",
       "\n",
       "                                          attritubes  \n",
       "0  gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  \n",
       "1  gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  \n",
       "2  gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  \n",
       "3  gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  \n",
       "4  gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  "
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#3.1 sort by start\n",
    "gtf.sort_values(by=['start'])\n",
    "gtf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "51fc227b-1a2b-4f76-b1b0-76e4be4a6a06",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['unknown'], dtype=object)"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#4. How many unique values for ‘source’ column\n",
    "gtf[\"source\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "005a44e2-f80b-40c6-a285-0491fc64687d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The unique catagory values for \"source\" column is: 1\n"
     ]
    }
   ],
   "source": [
    "#4.1 How many unique values for ‘source’ column\n",
    "gtf[\"source\"].nunique()\n",
    "print('The unique catagory values for \"source\" column is:', gtf[\"source\"].nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "46154e30-242d-4e8c-8773-21c3da39bd0b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attritubes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3214482</td>\n",
       "      <td>3216968</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>3216022</td>\n",
       "      <td>3216024</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>3216025</td>\n",
       "      <td>3216968</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>3421702</td>\n",
       "      <td>3421901</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3421702</td>\n",
       "      <td>3421901</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr   source     feature    start      end score strand frame  \\\n",
       "0  chr1  unknown        exon  3214482  3216968     .      -     .   \n",
       "1  chr1  unknown  stop_codon  3216022  3216024     .      -     .   \n",
       "2  chr1  unknown         CDS  3216025  3216968     .      -     2   \n",
       "3  chr1  unknown         CDS  3421702  3421901     .      -     1   \n",
       "4  chr1  unknown        exon  3421702  3421901     .      -     .   \n",
       "\n",
       "                                          attritubes  \n",
       "0  gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  \n",
       "1  gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  \n",
       "2  gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  \n",
       "3  gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  \n",
       "4  gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  "
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#5. Subset all values from ‘unknown’ and assign to new df called ‘ens_df’ \n",
    "mask = gtf[\"source\"] == \"unknown\"\n",
    "mask.to_frame()\n",
    "ens_df = gtf[mask]\n",
    "ens_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "bf015285-cb19-4fbe-ac3d-6c7fa85d2046",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attritubes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3214482</td>\n",
       "      <td>3216968</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>3216022</td>\n",
       "      <td>3216024</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>3216025</td>\n",
       "      <td>3216968</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>3421702</td>\n",
       "      <td>3421901</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3421702</td>\n",
       "      <td>3421901</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr   source     feature    start      end score strand frame  \\\n",
       "0  chr1  unknown        exon  3214482  3216968     .      -     .   \n",
       "1  chr1  unknown  stop_codon  3216022  3216024     .      -     .   \n",
       "2  chr1  unknown         CDS  3216025  3216968     .      -     2   \n",
       "3  chr1  unknown         CDS  3421702  3421901     .      -     1   \n",
       "4  chr1  unknown        exon  3421702  3421901     .      -     .   \n",
       "\n",
       "                                          attritubes  \n",
       "0  gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  \n",
       "1  gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  \n",
       "2  gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  \n",
       "3  gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  \n",
       "4  gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  "
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#5(alt.1). Subset all values from ‘unknown’ and assign to new df called ‘ens_df’ \n",
    "ens_df = gtf.query(\"source == 'unknown'\")\n",
    "ens_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ead0f937-54f1-4971-88dc-2cbc0740a395",
   "metadata": {},
   "outputs": [],
   "source": [
    "#5(alt.2). Subset all values from ‘unknown’ and assign to new df called ‘ens_df’ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "6f7023fc-da03-4293-95d0-5fe58a936f6b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attritubes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3214482</td>\n",
       "      <td>3216968</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>3216022</td>\n",
       "      <td>3216024</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>3216025</td>\n",
       "      <td>3216968</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>3421702</td>\n",
       "      <td>3421901</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3421702</td>\n",
       "      <td>3421901</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr   source     feature    start      end score strand frame  \\\n",
       "0  chr1  unknown        exon  3214482  3216968     .      -     .   \n",
       "1  chr1  unknown  stop_codon  3216022  3216024     .      -     .   \n",
       "2  chr1  unknown         CDS  3216025  3216968     .      -     2   \n",
       "3  chr1  unknown         CDS  3421702  3421901     .      -     1   \n",
       "4  chr1  unknown        exon  3421702  3421901     .      -     .   \n",
       "\n",
       "                                          attritubes  \n",
       "0  gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  \n",
       "1  gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  \n",
       "2  gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  \n",
       "3  gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  \n",
       "4  gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  "
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ens_df = gtf[gtf[\"source\"].str.contains(\"unknown\")]\n",
    "ens_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "2f6d576d-8e88-4d0c-a6d6-8b18a7def049",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "unknown    606183\n",
       "Name: source, dtype: int64"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#6. What is the size of ens_df? \n",
    "ens_df[\"source\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "4130923b-2bd6-42aa-8cc7-d51b991020ab",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The size of ends_df is (606183, 9)\n"
     ]
    }
   ],
   "source": [
    "#6(alt.1). What is the size of ens_df? \n",
    "size = ens_df.shape\n",
    "print ('The size of ends_df is', size)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "13ecef61-7ff9-477d-86e3-d3746e736042",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "feature      feature    \n",
       "CDS          CDS            261099\n",
       "exon         exon           290204\n",
       "start_codon  start_codon     27449\n",
       "stop_codon   stop_codon      27431\n",
       "Name: feature, dtype: int64"
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#7. Group ens_df by ‘feature’ column & apply the ‘count’ function to get the number of rows in each feature\n",
    "gb = ens_df.groupby(\"feature\")\n",
    "gb[\"feature\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "e4a05cc8-0632-431e-bd19-58ecd9696451",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attritubes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feature</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CDS</th>\n",
       "      <td>261099</td>\n",
       "      <td>261099</td>\n",
       "      <td>261099</td>\n",
       "      <td>261099</td>\n",
       "      <td>261099</td>\n",
       "      <td>261099</td>\n",
       "      <td>261099</td>\n",
       "      <td>261099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon</th>\n",
       "      <td>290204</td>\n",
       "      <td>290204</td>\n",
       "      <td>290204</td>\n",
       "      <td>290204</td>\n",
       "      <td>290204</td>\n",
       "      <td>290204</td>\n",
       "      <td>290204</td>\n",
       "      <td>290204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>start_codon</th>\n",
       "      <td>27449</td>\n",
       "      <td>27449</td>\n",
       "      <td>27449</td>\n",
       "      <td>27449</td>\n",
       "      <td>27449</td>\n",
       "      <td>27449</td>\n",
       "      <td>27449</td>\n",
       "      <td>27449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>stop_codon</th>\n",
       "      <td>27431</td>\n",
       "      <td>27431</td>\n",
       "      <td>27431</td>\n",
       "      <td>27431</td>\n",
       "      <td>27431</td>\n",
       "      <td>27431</td>\n",
       "      <td>27431</td>\n",
       "      <td>27431</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                chr  source   start     end   score  strand   frame  \\\n",
       "feature                                                               \n",
       "CDS          261099  261099  261099  261099  261099  261099  261099   \n",
       "exon         290204  290204  290204  290204  290204  290204  290204   \n",
       "start_codon   27449   27449   27449   27449   27449   27449   27449   \n",
       "stop_codon    27431   27431   27431   27431   27431   27431   27431   \n",
       "\n",
       "             attritubes  \n",
       "feature                  \n",
       "CDS              261099  \n",
       "exon             290204  \n",
       "start_codon       27449  \n",
       "stop_codon        27431  "
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#7(alt.1). Group ens_df by ‘feature’ column & apply the ‘count’ function to get the number of rows in each feature\n",
    "ens_df.groupby(\"feature\").count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "e78ef6ac-8712-42a6-b43f-3383b7eba575",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attritubes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3214482</td>\n",
       "      <td>3216968</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3421702</td>\n",
       "      <td>3421901</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3670552</td>\n",
       "      <td>3671498</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4290846</td>\n",
       "      <td>4293012</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Rp1\"; gene_name \"Rp1\"; p_id \"P16188\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4343507</td>\n",
       "      <td>4350091</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Rp1\"; gene_name \"Rp1\"; p_id \"P19224\";...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr   source feature    start      end score strand frame  \\\n",
       "0   chr1  unknown    exon  3214482  3216968     .      -     .   \n",
       "4   chr1  unknown    exon  3421702  3421901     .      -     .   \n",
       "6   chr1  unknown    exon  3670552  3671498     .      -     .   \n",
       "8   chr1  unknown    exon  4290846  4293012     .      -     .   \n",
       "11  chr1  unknown    exon  4343507  4350091     .      -     .   \n",
       "\n",
       "                                           attritubes  \n",
       "0   gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  \n",
       "4   gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  \n",
       "6   gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  \n",
       "8   gene_id \"Rp1\"; gene_name \"Rp1\"; p_id \"P16188\";...  \n",
       "11  gene_id \"Rp1\"; gene_name \"Rp1\"; p_id \"P19224\";...  "
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#8. Create a new df called ‘exons’ by subsetting all rows with ‘exon’ as a feature \n",
    "mask1 = gtf[\"feature\"] == \"exon\"\n",
    "mask1.to_frame()\n",
    "exons = gtf[mask1]\n",
    "exons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "91f398c7-9709-4b1e-b422-e8ce76bd5ac8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attritubes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3214482</td>\n",
       "      <td>3216968</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3421702</td>\n",
       "      <td>3421901</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3670552</td>\n",
       "      <td>3671498</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4290846</td>\n",
       "      <td>4293012</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Rp1\"; gene_name \"Rp1\"; p_id \"P16188\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4343507</td>\n",
       "      <td>4350091</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Rp1\"; gene_name \"Rp1\"; p_id \"P19224\";...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr   source feature    start      end score strand frame  \\\n",
       "0   chr1  unknown    exon  3214482  3216968     .      -     .   \n",
       "4   chr1  unknown    exon  3421702  3421901     .      -     .   \n",
       "6   chr1  unknown    exon  3670552  3671498     .      -     .   \n",
       "8   chr1  unknown    exon  4290846  4293012     .      -     .   \n",
       "11  chr1  unknown    exon  4343507  4350091     .      -     .   \n",
       "\n",
       "                                           attritubes  \n",
       "0   gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  \n",
       "4   gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  \n",
       "6   gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  \n",
       "8   gene_id \"Rp1\"; gene_name \"Rp1\"; p_id \"P16188\";...  \n",
       "11  gene_id \"Rp1\"; gene_name \"Rp1\"; p_id \"P19224\";...  "
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#8 (alt.1). Create a new df called ‘exons’ by subsetting all rows with ‘exon’ as a feature \n",
    "exon = gtf.query(\"feature == 'exon'\")\n",
    "exon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "645db1ec-bfc5-4062-88a7-f53aa8171855",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attritubes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3214482</td>\n",
       "      <td>3216968</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3421702</td>\n",
       "      <td>3421901</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3670552</td>\n",
       "      <td>3671498</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4290846</td>\n",
       "      <td>4293012</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Rp1\"; gene_name \"Rp1\"; p_id \"P16188\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4343507</td>\n",
       "      <td>4350091</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Rp1\"; gene_name \"Rp1\"; p_id \"P19224\";...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr   source feature    start      end score strand frame  \\\n",
       "0   chr1  unknown    exon  3214482  3216968     .      -     .   \n",
       "4   chr1  unknown    exon  3421702  3421901     .      -     .   \n",
       "6   chr1  unknown    exon  3670552  3671498     .      -     .   \n",
       "8   chr1  unknown    exon  4290846  4293012     .      -     .   \n",
       "11  chr1  unknown    exon  4343507  4350091     .      -     .   \n",
       "\n",
       "                                           attritubes  \n",
       "0   gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  \n",
       "4   gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  \n",
       "6   gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...  \n",
       "8   gene_id \"Rp1\"; gene_name \"Rp1\"; p_id \"P16188\";...  \n",
       "11  gene_id \"Rp1\"; gene_name \"Rp1\"; p_id \"P19224\";...  "
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#8 (alt.2). Create a new df called ‘exons’ by subsetting all rows with ‘exon’ as a feature\n",
    "exon = gtf[gtf[\"feature\"].str.contains(\"exon\")]\n",
    "exon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "13841a73-f535-4ce5-a1f1-83ad31a3a064",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/87/kfvv7k2j0dlb4jdfr8gtkc4c0000gn/T/ipykernel_915/2234964280.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  exons[\"gene_length\"] = exons[\"end\"] - exons[\"start\"]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attritubes</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3214482</td>\n",
       "      <td>3216968</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "      <td>2486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3421702</td>\n",
       "      <td>3421901</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "      <td>199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3670552</td>\n",
       "      <td>3671498</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "      <td>946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4290846</td>\n",
       "      <td>4293012</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Rp1\"; gene_name \"Rp1\"; p_id \"P16188\";...</td>\n",
       "      <td>2166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4343507</td>\n",
       "      <td>4350091</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Rp1\"; gene_name \"Rp1\"; p_id \"P19224\";...</td>\n",
       "      <td>6584</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr   source feature    start      end score strand frame  \\\n",
       "0   chr1  unknown    exon  3214482  3216968     .      -     .   \n",
       "4   chr1  unknown    exon  3421702  3421901     .      -     .   \n",
       "6   chr1  unknown    exon  3670552  3671498     .      -     .   \n",
       "8   chr1  unknown    exon  4290846  4293012     .      -     .   \n",
       "11  chr1  unknown    exon  4343507  4350091     .      -     .   \n",
       "\n",
       "                                           attritubes  gene_length  \n",
       "0   gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...         2486  \n",
       "4   gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...          199  \n",
       "6   gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...          946  \n",
       "8   gene_id \"Rp1\"; gene_name \"Rp1\"; p_id \"P16188\";...         2166  \n",
       "11  gene_id \"Rp1\"; gene_name \"Rp1\"; p_id \"P19224\";...         6584  "
      ]
     },
     "execution_count": 146,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#9. Create a new column called ‘gene_length’ and calculate the length of each gene\n",
    "exons[\"gene_length\"] = exons[\"end\"] - exons[\"start\"]\n",
    "exons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "f310550b-55f7-4219-ae89-f8b68006cb19",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attritubes</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3214482</td>\n",
       "      <td>3216968</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "      <td>2486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3421702</td>\n",
       "      <td>3421901</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "      <td>199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>3670552</td>\n",
       "      <td>3671498</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...</td>\n",
       "      <td>946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4290846</td>\n",
       "      <td>4293012</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Rp1\"; gene_name \"Rp1\"; p_id \"P16188\";...</td>\n",
       "      <td>2166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>4343507</td>\n",
       "      <td>4350091</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Rp1\"; gene_name \"Rp1\"; p_id \"P19224\";...</td>\n",
       "      <td>6584</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr   source feature    start      end score strand frame  \\\n",
       "0   chr1  unknown    exon  3214482  3216968     .      -     .   \n",
       "4   chr1  unknown    exon  3421702  3421901     .      -     .   \n",
       "6   chr1  unknown    exon  3670552  3671498     .      -     .   \n",
       "8   chr1  unknown    exon  4290846  4293012     .      -     .   \n",
       "11  chr1  unknown    exon  4343507  4350091     .      -     .   \n",
       "\n",
       "                                           attritubes  gene_length  \n",
       "0   gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...         2486  \n",
       "4   gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...          199  \n",
       "6   gene_id \"Xkr4\"; gene_name \"Xkr4\"; p_id \"P14345...          946  \n",
       "8   gene_id \"Rp1\"; gene_name \"Rp1\"; p_id \"P16188\";...         2166  \n",
       "11  gene_id \"Rp1\"; gene_name \"Rp1\"; p_id \"P19224\";...         6584  "
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#9(alt.1). Create a new column called ‘gene_length’ and calculate the length of each gene\n",
    "exons = exons.assign(gene_length = exons[\"end\"] - exons[\"start\"]) \n",
    "exons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "da71d95d-558a-4fb6-a86a-ccc304946f30",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>2.902040e+05</td>\n",
       "      <td>2.902040e+05</td>\n",
       "      <td>290204.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>7.561652e+07</td>\n",
       "      <td>7.561682e+07</td>\n",
       "      <td>297.704015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>4.448227e+07</td>\n",
       "      <td>4.448227e+07</td>\n",
       "      <td>655.025969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>4.240000e+02</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>3.589958e+07</td>\n",
       "      <td>3.590041e+07</td>\n",
       "      <td>93.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>7.435952e+07</td>\n",
       "      <td>7.435992e+07</td>\n",
       "      <td>133.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>1.094366e+08</td>\n",
       "      <td>1.094371e+08</td>\n",
       "      <td>197.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>1.951766e+08</td>\n",
       "      <td>1.951767e+08</td>\n",
       "      <td>83436.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              start           end    gene_length\n",
       "count  2.902040e+05  2.902040e+05  290204.000000\n",
       "mean   7.561652e+07  7.561682e+07     297.704015\n",
       "std    4.448227e+07  4.448227e+07     655.025969\n",
       "min    1.000000e+00  4.240000e+02       0.000000\n",
       "25%    3.589958e+07  3.590041e+07      93.000000\n",
       "50%    7.435952e+07  7.435992e+07     133.000000\n",
       "75%    1.094366e+08  1.094371e+08     197.000000\n",
       "max    1.951766e+08  1.951767e+08   83436.000000"
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#10 Calculate the mean and median of gene lengths \n",
    "exons.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de392c4d-6122-4cba-b70d-f0f5b28ef5bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#10.1 Calculate the mean and median of gene lengths "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "fe488dc5-9853-43e9-b590-65047262a06d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "297.70401510661463"
      ]
     },
     "execution_count": 157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons[\"gene_length\"].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4730d77-bcb7-4cb0-8455-c0ed1ce99807",
   "metadata": {},
   "outputs": [],
   "source": [
    "#10.2 Calculate the mean and median of gene lengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "796fa9bf-0bea-474d-abcc-1f8fa9ae8540",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "133.0"
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons[\"gene_length\"].median()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d543f51-fd3f-411e-be9e-404ddc545165",
   "metadata": {},
   "outputs": [],
   "source": [
    "#11. Find the top 10 longest exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef4683e6-52f4-45f7-836f-2cb0aefa9289",
   "metadata": {},
   "outputs": [],
   "source": [
    "exons.sort_values(by=['gene_length'], ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6bfa1cc0-5e30-4514-8805-6f5e78ee8ecd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#12. Find the lowest 20 shorest exons "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3cbf95e",
   "metadata": {},
   "outputs": [],
   "source": [
    "exons.sort_values(by=['gene_length']).head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "116eb027-0ec1-453c-9832-85668b3049e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#13. Split attributes column on ‘;’ and create new dataframe, merge this onto the ens_gtf dataframe using pandas merge "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0d72dfc",
   "metadata": {},
   "outputs": [],
   "source": [
    "exons2=exons[\"attribute\"].str.split(\";\", expand=True)\n",
    "exons2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58d2801b",
   "metadata": {},
   "outputs": [],
   "source": [
    "exons3 = exons.join(exons2)\n",
    "exons3.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "899c2ae9-9e56-487b-9cee-271300f1cda6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Optional. Count the number of genes per chromosome, and add the number of genes per chomosome onto the idx_df from earlie"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
